dropshot_endpoint/src/lib.rs

```diff
@@ -167,53 +195,21 @@ fn do_endpoint(
                     let span = Error::new_spanned(pat.ty.as_ref(), "").span();
                     let ty = pat.ty.as_ref().into_token_stream();
 
-                    if i == 0 {
-                        quote_spanned! { span=>
-                            const _: fn() = ||{
-                                trait TypeEq {
-                                    type This: ?Sized;
-                                }
-                                impl<T: ?Sized> TypeEq for T {
-                                    type This = Self;
-                                }
-                                fn need_arc_requestcontext<T>()
-                                where
-                                    T: ?Sized + TypeEq<This = #req>,
-                                {
-                                }
-                                need_arc_requestcontext::<#ty>();
```

Comment:
do we have a test that skips the request context and just has extractors?

Reply:
Oooooo apparently not, but I'll add one (I'm going to add it to `tests/fail/bad_endpoint...` since those handle output a little better for failure cases which occur after the macro finishes processing successfully).

The output is currently:

```
error[E0277]: the trait bound `dropshot::Query<QueryParams>: RequestContextArgument` is not satisfied
  --> $DIR/bad_endpoint9.rs:21:1
   |
21 | / #[endpoint {
22 | |     method = GET,
23 | |     path = "/test",
24 | | }]
   | |__^ the trait `RequestContextArgument` is not implemented for `dropshot::Query<QueryParams>`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I don't *think* it's possible to catch this error within the `endpoint` macro, as it relies on type information for the first argument.

I updated the `RequestContextArgument` documentation to reiterate expected usage. If you don't think that's sufficient, I could look into emitting more structures-that-act-as-checks from the macro?

Reply:
If I understand you correctly (and I'm still kind of paging this back in), I believe the code that's deleted here was intended to test for type equality of that first argument. If for some reason this check no longer works with the more complex RequestContext type I think it's fine to leave as is.

... I did think this was a clever solution to checking for type equality, but cleverness alone isn't reason enough to keep it ;-)

Reply:
Yeah, the problem is that the `req` parameter in the old version was hardcoded as `std::sync::Arc<RequestContext>`, but we don't actually know the intended type this time, as it has a nested generic. Previously, the code emitted a (if I'm reading this correctly) constant lambda filled with a trait, trait implementation, and function requiring that trait implementation, all to backpropagate errors about type equality. If we don't know the exact type we need, I'm a little worried the propagation of an error through this mechanism might deliver more confusion than clarity (since it's all to ask for a type that we can't know for certain), and I'm not sure of the best way to adapt it.

Besides, the current error (IMO) seems to provide just as much value, saying "the first argument needs to be a `RequestContextArgument`", which explains the problem, and elaborates in the documentation of `RequestContextArgument`.